radicle/src/git.rs

```diff
@@ -104,27 +119,34 @@ pub fn remote_refs(url: &Url) -> Result<HashMap<RemoteId, Refs>, ListRefsError>
         let (id, refname) = parse_ref::<PublicKey>(r.name())?;
         let entry = remotes.entry(id).or_insert_with(Refs::default);
 
-        entry.insert(refname, r.oid().into());
+        entry.insert(refname.into(), r.oid().into());
     }
 
     Ok(remotes)
 }
 
 /// Parse a ref string.
-pub fn parse_ref<T: FromStr>(s: &str) -> Result<(T, format::RefString), RefError> {
+pub fn parse_ref<T>(s: &str) -> Result<(T, format::Qualified), RefError>
+where
+    T: FromStr,
+    T::Err: std::error::Error + Send + Sync + 'static,
+{
     let input = format::RefStr::try_from_str(s)?;
-    let suffix = input
-        .strip_prefix(format::refname!("refs/remotes"))
-        .ok_or_else(|| RefError::InvalidName(input.to_owned()))?;
-
-    let mut components = suffix.components();
-    let id = components
-        .next()
-        .ok_or_else(|| RefError::InvalidName(input.to_owned()))?;
-    let id = T::from_str(&id.to_string()).map_err(|_| RefError::InvalidName(input.to_owned()))?;
-    let refstr = components.collect::<format::RefString>();
-
-    Ok((id, refstr))
+    match input.namespaced() {
```

Comment:
Meta: I find this API or naming strange: it sounds like it is namespacing `input`, ie. "give me a namespaced version of input", even though input is already namespaced. I think perhaps this should be called `as_namespaced()` or `Namespaced::try_from(input)`